local-cli/runAndroid/runAndroid.js

```diff
@@ -78,13 +78,108 @@ function tryRunAdbReverse(device) {
+      'app/src/main/AndroidManifest.xml',
+      'utf8'
+    ).match(/package="(.+?)"/)[1];
+
+  const adbPath = getAdbPath();
+  if (args.deviceId) {
+    runOnSpecificDevice(args, cmd, packageName, adbPath);
+  } else {
+    runOnAllDevices(args, cmd, packageName, adbPath);
+  }
+}
+
+function runOnSpecificDevice(args, cmd, packageName, adbPath) {
+  let devices = adb.getDevices();
+  let selectedDevice = findMatchingDevice(devices, args.deviceId);
+  if (selectedDevice){
+    buildApk(cmd);
+    installAndLaunchOnDevice(args, selectedDevice, packageName, adbPath);
+  } else {
+      if (devices){
+        console.log('Could not find device with the id: "' + args.deviceId + '".');
+        console.log('Choose one of the following:');
+        child_process.execFileSync('adb', ['devices'], {
+          stdio: [process.stdin, process.stdout, process.stderr],
+        });
+      } else {
+        console.log('No android devices connected.');
+      }
+  }
+}
+
+function buildApk(cmd) {
   try {
-    adb.getDevices().map((device) => tryRunAdbReverse(device));
+    console.log(chalk.bold(
+      'Building the app...'
+    ));
+
+    child_process.execFileSync(cmd, ['build', '-x', 'lint'], {
+      stdio: [process.stdin, process.stdout, process.stderr],
+    });
+  } catch (e) {
+    console.log(chalk.red(
+      'Could not build the app on the device, read the error above for details.\n'
+    ));
+  }
+}
 
-    const cmd = process.platform.startsWith('win')
-      ? 'gradlew.bat'
-      : './gradlew';
+function tryInstallAppOnDevice(args, device) {
+  try {
+    const pathToApk = 'app/build/outputs/apk/app-debug.apk';
+    const adbPath = getAdbPath();
+    const adbArgs = ['-s', device, 'install', pathToApk];
+    console.log(chalk.bold(
+      `Installing the app on the device (cd android && adb -s ${device} install ${pathToApk}`
+    ));
+    child_process.execFileSync(adbPath, adbArgs, {
+      stdio: [process.stdin, process.stdout, process.stderr],
+    });
+  } catch (e) {
+    console.log(e.message);
+    console.log(chalk.red(
+      'Could not install the app on the device, read the error above for details.\n'
+    ));
+  }
+}
 
+function tryLaunchAppOnDevice(device, packageName, adbPath) {
+  try {
+    const adbArgs = ['-s', device, 'shell', 'am', 'start', '-n', packageName + '/.MainActivity'];
+    console.log(chalk.bold(
+      `Starting the app on ${device} (${adbPath} ${adbArgs.join(' ')})...`
+    ));
+    child_process.spawnSync(adbPath, adbArgs, {stdio: 'inherit'});
+  } catch (e) {
+    console.log(chalk.red(
+      'adb invocation failed. Do you have adb in your PATH?'
+    ));
+  }
+}
+
+function installAndLaunchOnDevice(args, selectedDevice, packageName, adbPath) {
+  tryRunAdbReverse(selectedDevice);
+  tryInstallAppOnDevice(args, selectedDevice);
+  tryLaunchAppOnDevice(selectedDevice, packageName, adbPath);
+}
+
+function findMatchingDevice(devices, id) {
+  for (let i = devices.length - 1; i >= 0; i--) {
```

Comment:
Just curious - why search from the end? How about using `filter` to make this shorter?